### PR TITLE
refactor: Use `write_fledgling()` to write `NEWS.md`

### LINKED ISSUE #147: Collect in-dev items under “development version” heading
In between releases, it'd be great if `update_news()` could collect the draft news items under a specific heading. My personal preference would be to follow [tidyverse conventions](https://style.tidyverse.org/news.html#development):

```
# <package> (development version)
```

One thing I like about the above heading, as opposed to using a `.9000` version suffix, is that it allows the package maintainer some freedom if the package version changes without being officially released, while making it easier to automate maintenance of the NEWS file between official releases.

### LINKED ISSUE #336: Sort NEWS.md entries according to conventional commits headers

### LINKED ISSUE #586: use simple fledgling for writing operations

### PR DESCRIPTION
**Pull request by maelle**
Fix #147
Fix #586 
Fix #336 

TODOs

- [x] decide whether to deprecate `update_version()`.
- [x] decide what to name `update_news()` if not `update_news()`. I am not against keeping the name. I've thought of `record_changes()`.
- [ ] would there be a way to first test this branch in a few repos where fledge is used?
- [ ] improve the section header parsing (see https://github.com/cynkra/fledge/pull/588#pullrequestreview-1241929643 or keep it as a debt-labelled issue if we plan a debt-paying-back session soon? :sweat_smile: )

### CODE REVIEW

**Comment by maelle** [comment]
with `update_news()` calling `update_version()`, I wonder whether it should be called `update_version()` :thinking:

**Comment by maelle** [comment]
``` r
fledge:::parse_news_md()
#> $`fledge 0.1.0.9025 (2022-11-19)`
#> $`fledge 0.1.0.9025 (2022-11-19)`[[1]]
#> $`fledge 0.1.0.9025 (2022-11-19)`[[1]]$Documentation
#> $`fledge 0.1.0.9025 (2022-11-19)`[[1]]$Documentation[[1]]
#> [1] "-   Typo."
#> 
#> 
#> 
#> 
#> $`fledge 0.1.0.9024 (2022-11-12)`
#> $`fledge 0.1.0.9024 (2022-11-12)`[[1]]
#> [1] "-   Internal changes only."
#> 
#> 
#> $`fledge 0.1.0.9023 (2022-11-11)`
#> $`fledge 0.1.0.9023 (2022-11-11)`[[1]]
#> $`fledge 0.1.0.9023 (2022-11-11)`[[1]]$`Bug fixes`
#> $`fledge 0.1.0.9023 (2022-11-11)`[[1]]$`Bug fixes`[[1]]
#> [1] "-   Remove partial matching warning (#579)."
#> 
#> 
#> 
#> $`fledge 0.1.0.9023 (2022-11-11)`[[2]]
#> $`fledge 0.1.0.9023 (2022-11-11)`[[2]]$Features
#> $`fledge 0.1.0.9023 (2022-11-11)`[[2]]$Features[[1]]
#> [1] "-   Draft fledgeling object, ready for internal use (#581)."
#> [2] ""                                                           
#> [3] "-   Add full stop to messages (#578)."                      
#> 
#> 
#> 
#> $`fledge 0.1.0.9023 (2022-11-11)`[[3]]
#> $`fledge 0.1.0.9023 (2022-11-11)`[[3]]$Chore
#> $`fledge 0.1.0.9023 (2022-11-11)`[[3]]$Chore[[1]]
#> [1] "-   Store downlit info (#580)."
#> 
#> 
#> 
#> $`fledge 0.1.0.9023 (2022-11-11)`[[4]]
#> $`fledge 0.1.0.9023 (2022-11-11)`[[4]]$Documentation
#> $`fledge 0.1.0.9023 (2022-11-11)`[[4]]$Documentation[[1]]
#> [1] "-   Cynkratemplate::use_cynkra_pkgdown() + re-add comment"                
#> [2] "    ([\\@krlmlr]{.citation cites=\"krlmlr\"}, [\\@41898282+github-actions"
#> [3] "    \\[bot\\]]{.citation cites=\"41898282+github-actions\"},"             
#> [4] "    [\\@maelle]{.citation cites=\"maelle\"}, #568)."                      
#> 
#> 
#> 
#> $`fledge 0.1.0.9023 (2022-11-11)`[[5]]
#> $`fledge 0.1.0.9023 (2022-11-11)`[[5]]$Uncategorized
#> $`fledge 0.1.0.9023 (2022-11-11)`[[5]]$Uncategorized[[1]]
#> [1] "-   Split news (#584)."
#> 
#> 
#> 
#> 
#> $`fledge 0.1.0.9022 (2022-11-10)`
#> $`fledge 0.1.0.9022 (2022-11-10)`[[1]]
#> $`fledge 0.1.0.9022 (2022-11-10)`[[1]]$`Bug fixes`
#> $`fledge 0.1.0.9022 (2022-11-10)`[[1]]$`Bug fixes`[[1]]
#> [1] "-   Add full stop for entries"
#> 
#> 
#> 
#> 
#> $`fledge 0.1.0.9021 (2022-11-09)`
#> $`fledge 0.1.0.9021 (2022-11-09)`[[1]]
#> [1] "-   Internal changes only."
#> 
#> 
#> $`fledge 0.1.0.9020 (2022-11-03)`
#> $`fledge 0.1.0.9020 (2022-11-03)`[[1]]
#> [1] "-   Snapshot updates for R-CMD-check-base (null) (#570)"
#> 
#> 
#> $`fledge 0.1.0.9019 (2022-10-28)`
#> $`fledge 0.1.0.9019 (2022-10-28)`[[1]]
#> [1] "-   Internal changes only."
#> 
#> 
#> $`fledge 0.1.0.9018 (2022-10-26)`
#> $`fledge 0.1.0.9018 (2022-10-26)`[[1]]
#> $`fledge 0.1.0.9018 (2022-10-26)`[[1]]$Documentation
#> $`fledge 0.1.0.9018 (2022-10-26)`[[1]]$Documentation[[1]]
#> [1] "-   Fix URL (#573)"
#> 
#> 
#> 
#> 
#> $`fledge 0.1.0.9017 (2022-10-24)`
#> $`fledge 0.1.0.9017 (2022-10-24)`[[1]]
#> $`fledge 0.1.0.9017 (2022-10-24)`[[1]]$`Bug fixes`
#> $`fledge 0.1.0.9017 (2022-10-24)`[[1]]$`Bug fixes`[[1]]
#> [1] "-   Relax requirement for GitHub PAT scope (#572)"
#> 
#> 
#> 
#> 
#> $`fledge 0.1.0.9016 (2022-10-20)`
#> $`fledge 0.1.0.9016 (2022-10-20)`[[1]]
#> [1] "-   Internal changes only."
#> 
#> 
#> $`fledge 0.1.0.9015 (2022-10-19)`
#> $`fledge 0.1.0.9015 (2022-10-19)`[[1]]
#> [1] "-   Internal changes only."
#> 
#> 
#> $`fledge 0.1.0.9014 (2022-10-18)`
#> $`fledge 0.1.0.9014 (2022-10-18)`[[1]]
#> [1] "-   Harmonize yaml formatting"       
#> [2] ""                                    
#> [3] "-   Revert changes to matrix section"
#> 
#> 
#> $`fledge 0.1.0.9013 (2022-10-16)`
#> $`fledge 0.1.0.9013 (2022-10-16)`[[1]]
#> [1] "-   Internal changes only."
#> 
#> 
#> $`fledge 0.1.0.9012 (2022-10-14)`
#> $`fledge 0.1.0.9012 (2022-10-14)`[[1]]
#> [1] "-   Internal changes only."
#> 
#> 
#> $`fledge 0.1.0.9011 (2022-10-11)`
#> $`fledge 0.1.0.9011 (2022-10-11)`[[1]]
#> [1] "-   Internal changes only."
#> 
#> 
#> $`fledge 0.1.0.9010 (2022-09-20)`
#> $`fledge 0.1.0.9010 (2022-09-20)`[[1]]
#> $`fledge 0.1.0.9010 (2022-09-20)`[[1]]$Features
#> $`fledge 0.1.0.9010 (2022-09-20)`[[1]]$Features[[1]]
#> [1] "-   `bump_version()` returns `TRUE` if a new version has been created"
#> 
#> 
#> 
#> 
#> $`fledge 0.1.0.9009 (2022-09-20)`
#> $`fledge 0.1.0.9009 (2022-09-20)`[[1]]
#> $`fledge 0.1.0.9009 (2022-09-20)`[[1]]$Features
#> $`fledge 0.1.0.9009 (2022-09-20)`[[1]]$Features[[1]]
#> [1] "-   Better support for merge commits of all kinds ([\\@maelle]{.citation"
#> [2] "    cites=\"maelle\"}, #512)"                                            
#> 
#> 
#> 
#> $`fledge 0.1.0.9009 (2022-09-20)`[[2]]
#> $`fledge 0.1.0.9009 (2022-09-20)`[[2]]$Uncategorized
#> $`fledge 0.1.0.9009 (2022-09-20)`[[2]]$Uncategorized[[1]]
#> [1] "-   Internal changes only."
#> 
#> 
#> 
#> 
#> $`fledge 0.1.0.9008 (2022-09-18)`
#> $`fledge 0.1.0.9008 (2022-09-18)`[[1]]
#> $`fledge 0.1.0.9008 (2022-09-18)`[[1]]$Features
#> $`fledge 0.1.0.9008 (2022-09-18)`[[1]]$Features[[1]]
#> [1] "-   Better support for merge commits of all kinds ([\\@maelle]{.citation"
#> [2] "    cites=\"maelle\"}, #512)"                                            
#> 
#> 
#> 
#> $`fledge 0.1.0.9008 (2022-09-18)`[[2]]
#> $`fledge 0.1.0.9008 (2022-09-18)`[[2]]$Uncategorized
#> $`fledge 0.1.0.9008 (2022-09-18)`[[2]]$Uncategorized[[1]]
#> [1] "-   Internal changes only."
#> 
#> 
#> 
#> 
#> $`fledge 0.1.0.9007 (2022-09-16)`
#> $`fledge 0.1.0.9007 (2022-09-16)`[[1]]
#> $`fledge 0.1.0.9007 (2022-09-16)`[[1]]$Features
#> $`fledge 0.1.0.9007 (2022-09-16)`[[1]]$Features[[1]]
#> [1] "-   Better support for merge commits of all kinds ([\\@maelle]{.citation"
#> [2] "    cites=\"maelle\"}, #512)"                                            
#> 
#> 
#> 
#> $`fledge 0.1.0.9007 (2022-09-16)`[[2]]
#> $`fledge 0.1.0.9007 (2022-09-16)`[[2]]$Uncategorized
#> $`fledge 0.1.0.9007 (2022-09-16)`[[2]]$Uncategorized[[1]]
#> [1] "-   Internal changes only."
#> 
#> 
#> 
#> 
#> $`fledge 0.1.0.9006 (2022-09-14)`
#> $`fledge 0.1.0.9006 (2022-09-14)`[[1]]
#> $`fledge 0.1.0.9006 (2022-09-14)`[[1]]$Features
#> $`fledge 0.1.0.9006 (2022-09-14)`[[1]]$Features[[1]]
#> [1] "-   Better support for merge commits of all kinds ([\\@maelle]{.citation"
#> [2] "    cites=\"maelle\"}, #512)"                                            
#> 
#> 
#> 
#> $`fledge 0.1.0.9006 (2022-09-14)`[[2]]
#> $`fledge 0.1.0.9006 (2022-09-14)`[[2]]$Uncategorized
#> $`fledge 0.1.0.9006 (2022-09-14)`[[2]]$Uncategorized[[1]]
#> [1] "-   Internal changes only."
#> 
#> 
#> 
#> 
#> $`fledge 0.1.0.9005 (2022-09-13)`
#> $`fledge 0.1.0.9005 (2022-09-13)`[[1]]
#> [1] "-   Internal changes only."
#> 
#> 
#> $`fledge 0.1.0.9004 (2022-09-12)`
#> $`fledge 0.1.0.9004 (2022-09-12)`[[1]]
#> [1] "-   Internal changes only."
#> 
#> 
#> $`fledge 0.1.0.9003 (2022-09-12)`
#> $`fledge 0.1.0.9003 (2022-09-12)`[[1]]
#> $`fledge 0.1.0.9003 (2022-09-12)`[[1]]$`Bug fixes`
#> $`fledge 0.1.0.9003 (2022-09-12)`[[1]]$`Bug fixes`[[1]]
#> [1] "-   Adjust empty lines and space in update_news() (#357)"
#> [2] ""                                                        
#> [3] "-   Add PR number to CC messages (#353)."                
#> [4] ""                                                        
#> [5] "-   Fix conventional commit regex (#329)"                
#> 
#> 
#> 
#> $`fledge 0.1.0.9003 (2022-09-12)`[[2]]
#> $`fledge 0.1.0.9003 (2022-09-12)`[[2]]$Features
#> $`fledge 0.1.0.9003 (2022-09-12)`[[2]]$Features[[1]]
#>  [1] "-   New default `pre_release(which = \"next\")` (#521, #524)."            
#>  [2] ""                                                                         
#>  [3] "-   Add which \"pre-minor\" and \"pre-major\" to update_version() (#413," 
#>  [4] "    #412)."                                                               
#>  [5] ""                                                                         
#>  [6] "-   Include reference to issue closed with a PR (#361, #411)."            
#>  [7] ""                                                                         
#>  [8] "-   Capitalize first letter of each bullet (#360, #409)."                 
#>  [9] ""                                                                         
#> [10] "-   Extract contributor name from PR merge message ([\\@krlmlr]{.citation"
#> [11] "    cites=\"krlmlr\"}, #358)"                                             
#> [12] ""                                                                         
#> [13] "-   Harvest PR attribution from squash commits ([\\@maelle]{.citation"    
#> [14] "    cites=\"maelle\"}, #349)"                                             
#> [15] ""                                                                         
#> [16] "-   Harvest PR title from merge commit messages ([\\@krlmlr]{.citation"   
#> [17] "    cites=\"krlmlr\"}, [\\@maelle]{.citation cites=\"maelle\"}, #343)"    
#> [18] ""                                                                         
#> [19] "-   Improve parsing of conventional commit messages (#332)."              
#> [20] ""                                                                         
#> [21] "-   Improve bump_version() (error) messages ([\\@maelle]{.citation"       
#> [22] "    cites=\"maelle\"}, #328)"                                             
#> [23] ""                                                                         
#> [24] "-   Improve bump_version() behavior in the absence of changes"            
#> [25] "    ([\\@krlmlr]{.citation cites=\"krlmlr\"}, [\\@maelle]{.citation"      
#> [26] "    cites=\"maelle\"}, #323)"                                             
#> [27] ""                                                                         
#> [28] "-   New `local_test_project()` (#318)"                                    
#> [29] ""                                                                         
#> [30] "-   Extract conventional commit messages for the changelog (#314)."       
#> 
#> 
#> 
#> $`fledge 0.1.0.9003 (2022-09-12)`[[3]]
#> $`fledge 0.1.0.9003 (2022-09-12)`[[3]]$Chore
#> $`fledge 0.1.0.9003 (2022-09-12)`[[3]]$Chore[[1]]
#> [1] "-   Turn Netlify builds off for now (#326)."             
#> [2] ""                                                        
#> [3] "-   Enable auto-style on GitHub Actions (#317)."         
#> [4] ""                                                        
#> [5] "-   Remove testthat specialization for snapshots (#309)."
#> 
#> 
#> 
#> $`fledge 0.1.0.9003 (2022-09-12)`[[4]]
#> $`fledge 0.1.0.9003 (2022-09-12)`[[4]]$`Continuous integration`
#> $`fledge 0.1.0.9003 (2022-09-12)`[[4]]$`Continuous integration`[[1]]
#> [1] "-   Create fledge.yaml (#520)."
#> 
#> 
#> 
#> $`fledge 0.1.0.9003 (2022-09-12)`[[5]]
#> $`fledge 0.1.0.9003 (2022-09-12)`[[5]]$Documentation
#> $`fledge 0.1.0.9003 (2022-09-12)`[[5]]$Documentation[[1]]
#> [1] "-   Add README section on related tools (#527)."                    
#> [2] ""                                                                   
#> [3] "-   Update summary of how fledge uses commit messages (#499, #511)."
#> [4] ""                                                                   
#> [5] "-   Fix typo in README (#501)"                                      
#> 
#> 
#> 
#> $`fledge 0.1.0.9003 (2022-09-12)`[[6]]
#> $`fledge 0.1.0.9003 (2022-09-12)`[[6]]$Uncategorized
#> $`fledge 0.1.0.9003 (2022-09-12)`[[6]]$Uncategorized[[1]]
#> [1] "-   Create and open draft release directly, without using"    
#> [2] "    `usethis::use_github_release()`."                         
#> [3] ""                                                             
#> [4] "-   Create tag as part of `release()`."                       
#> [5] ""                                                             
#> [6] "-   Fix `post_release()`, still need to tag released version."
#> 
#> 
#> 
#> 
#> $`fledge 0.1.0.9002 (2022-04-02)`
#> $`fledge 0.1.0.9002 (2022-04-02)`[[1]]
#> [1] "-   `release()` no longer asks for confirmation."                     
#> [2] "-   Inline `devtools::submit_cran()` minus the confirmation messages."
#> 
#> 
#> $`fledge 0.1.0.9001 (2022-02-22)`
#> $`fledge 0.1.0.9001 (2022-02-22)`[[1]]
#> [1] "-   New `pre_release()`, `release()` and `post_release()` for"
#> [2] "    semi-automatic CRAN releases (#27, #28)."                 
#> [3] "-   Separate snapshot tests for dev version of testthat."     
#> 
#> 
#> $`fledge 0.1.0.9000 (2021-12-07)`
#> $`fledge 0.1.0.9000 (2021-12-07)`[[1]]
#> [1] "-   Same as previous version."
#> 
#> 
#> $`fledge 0.1.0 (2021-12-07)`
#> $`fledge 0.1.0 (2021-12-07)`[[1]]
#>  [1] "Change log management utility, initial release."                          
#>  [2] ""                                                                         
#>  [3] "The main entry point is `bump_version()`, which does the following:"      
#>  [4] ""                                                                         
#>  [5] "1.  Checks that `DESCRIPTION` and `NEWS.md` are clean before bumping,"    
#>  [6] "    and that the current branch is the main branch."                      
#>  [7] "2.  `update_news()`: collects `NEWS` entries from top-level commits"      
#>  [8] "3.  `update_version()`: bump version in `DESCRIPTION`, add header to"     
#>  [9] "    `NEWS.md`"                                                            
#> [10] "4.  Depending on the kind of update:"                                     
#> [11] "    -   If \"dev\", `finalize_version()`: commit `DESCRIPTION` and"       
#> [12] "        `NEWS.md`, create tag with message"                               
#> [13] "    -   Otherwise, `commit_version()`; the user needs to call"            
#> [14] "        `tag_version()` manually"                                         
#> [15] "5.  Suggests to push the changes if an upstream repository is"            
#> [16] "    configured."                                                          
#> [17] ""                                                                         
#> [18] "If you haven't committed since updating `NEWS.md` and `DESCRIPTION`, you" 
#> [19] "can also edit `NEWS.md` and call `tag_version()` again. Both the commit"  
#> [20] "and the tag will be updated."                                             
#> [21] ""                                                                         
#> [22] "Bumping can be undone with `unbump_version()`."                           
#> [23] ""                                                                         
#> [24] "If the `DESCRIPTION` has a `\"Date\"` field, it is populated with the"    
#> [25] "current date in `update_version()`. Likewise, if `NEWS.md` contains"      
#> [26] "dates in the headers, new versions also get a date."                      
#> [27] ""                                                                         
#> [28] "An empty list of changes adds a \"Same as previous version\" bullet. This"
#> [29] "allows bumping to a dev version immediately after CRAN release."          
#> [30] ""                                                                         
#> [31] "Also includes helper functions `get_last_tag()` and"                      
#> [32] "`get_top_level_commits()`."                                               
#> [33] ""                                                                         
#> [34] "Includes vignettes: \"Get started\", \"Using fledge\", and \"Fledge"      
#> [35] "internals\". Examples and tests are supported with a demo project,"       
#> [36] "created via `with_demo_project()`."                                       
#> [37] ""                                                                         
#> [38] "Thanks Patrick Schratz and Maëlle Salmon for your contributions!"
```

<sup>Created on 2022-11-21 with [reprex v2.0.2](https://reprex.tidyverse.org)</sup>

**Comment by maelle** [comment]
In `parse_news_md()`, maybe already parse a bit further, so that all versions have the same structure. A list of lists, one list per present "type" in the version, items with no type under "Uncategorized". When writing, only add the "Uncategorized" header if there are other types present in the version.

File: NEWS.md [hunk 1]
```diff
@@ -11,7 +11,6 @@
 
 - Internal changes only.
 
-
```
Comment (krlmlr) [inline NEWS.md:14]: I want an empty line here, contrary to Markdown style guides.

File: vignettes/internals.Rmd [hunk 2]
```diff
@@ -62,9 +62,7 @@ You can use second and third level headers and add as many empty lines as you wa
 
 The first function to call at any stage is always `bump_version()`. This does the following
 
-1. Calls `update_news()` to collect NEWS entries from top-level commits.
-
-2. Calls `update_version()` to increment the version in `DESCRIPTION` and add a header to `NEWS.md`.
+1. Calls `update_news()` to collect NEWS entries from top-level commits and to increment the version in `DESCRIPTION` and add a header to `NEWS.md`.
```
Comment (krlmlr) [inline vignettes/internals.Rmd:65]: Need a new name here, and deprecate the existing?

Comment (maelle) [reply vignettes/internals.Rmd:65]: receipt

**Comment by maelle** [comment]
Realizing the nested list really needs to be the simplest one possible, to simplify handling it when regrouping.

**Comment by maelle** [comment]
So TODOS are

- [ ] edit the parsing so that every version is a consistent list of lists including empty ones.
- [ ] then use that in the regrouping.

**Comment by maelle** [comment]
thought, after fighting with line wrapping: when writing we might want to only write the lines changed in the fledgeling object (probably only something at the top, we could ignore other cases for now).

File: R/pandoc.R [hunk 3]
```diff
@@ -0,0 +1,98 @@
+parse_news_md <- function(news = brio::read_lines(news_path())) {
```
Comment (maelle) [inline R/pandoc.R:1]: so we're basically writing a parser for NEWS file...

**Comment by maelle** [comment]
New TODOS

- [ ] make it work for subsections in types!
- [ ] add regrouping test

File: R/fledgling.R [hunk 4]
```diff
@@ -96,3 +137,84 @@ trim_empty_lines <- function(x) {
...
+paste_news_lines <- function(lines, header_level) {
+  lines <- unlist(lines, recursive = FALSE)
+  if (rlang::is_named(lines)) {
+    header_sign <- paste(rep("#", header_level), collapse = "")
+    lines <- purrr::imap_chr(
+      lines,
+      ~ sprintf(
+        "%s %s\n\n%s",
+        header_sign,
+        .y, paste_news_lines(.x,
```
Comment (maelle) [inline R/fledgling.R:211]: recursive

File: NEWS.md (hunk 1 above)
Comment (maelle) [reply NEWS.md:14]: Added :heavy_check_mark:

File: NAMESPACE [hunk 5]
```diff
@@ -13,7 +13,6 @@ export(release)
 export(tag_version)
 export(unbump_version)
 export(update_news)
-export(update_version)
```
Comment (maelle) [inline NAMESPACE:16]: if we agree on the strategy, this should be deprecated, not just removed :innocent:

File: NEWS.md [hunk 6]
```diff
@@ -30,15 +29,14 @@
 
 ## Documentation
 
-- Cynkratemplate::use_cynkra_pkgdown() + re-add comment (@krlmlr, @41898282+github-actions[bot], @maelle, #568).
+- Cynkratemplate::use_cynkra_pkgdown() + re-add comment (@krlmlr,
```
Comment (maelle) [inline NEWS.md:32]: cf #592

File: R/pandoc.R [hunk 7]
```diff
@@ -0,0 +1,101 @@
...
+  pandoc::pandoc_run(c("-t", "html", "--wrap=preserve", "-f", "gfm", "-o", out_temp_file, temp_file, "--section-divs"))
+
+  html <- xml2::read_html(out_temp_file, encoding = "UTF-8")
+
+  if (length(xml2::xml_contents(html)) == 0) {
+    return(NULL)
+  }
+
+  version_header_level <- 1
+  versions <- xml2::xml_find_all(html, ".//section[@class='level1']")
```
Comment (maelle) [inline R/pandoc.R:17]: having the section **divs** helps a lot with parsing

**Review by krlmlr — COMMENTED** [review]
How can we make wrapping of the `NEWS.md` optional?

**Comment by maelle** [comment]
> How can we make wrapping of the NEWS.md optional?

What do you mean by wrapping?

**Comment by maelle** [comment]
Do you mean _line_ wrapping? If so see https://github.com/cynkra/fledge/pull/588#issuecomment-1333714175

**Comment by krlmlr** [comment]
Does this mean that we can't really opt out of wrapping lines in `NEWS.md` ?

**Comment by maelle** [comment]
I can explore more, I remember trying many Pandoc options but I haven't kept a log which was not smart.

**Comment by krlmlr** [comment]
If pandoc is difficult, can we avoid it?

**Comment by maelle** [comment]
We need to parse NEWS, and any parsing method (Pandoc, commonmark/xslt via tinkr, regular expressions, etc.) will bring its own challenges.

**Comment by maelle** [comment]
The wrapping now looks better :woman_shrugging:

**Comment by krlmlr** [comment]
Can you please resolve conflicts?

**Comment by krlmlr** [comment]
This PR is now slightly too large, can we move the change to the fledge URL to a separate PR?

**Comment by krlmlr** [comment]
It's great we got the wrapping right. I want to understand what's happening.

**Comment by maelle** [comment]
> Can you please resolve conflicts?

Done!

> This PR is now slightly too large, can we move the change to the fledge URL to a separate PR?

I'm sorry, I don't understand what change you mean. I see no relevant change in `DESCRIPTION`/`_pkgdown.yml`.

> It's great we got the wrapping right. I want to understand what's happening.

What's needed from me?

File: vignettes/internals.Rmd (hunk 2 above)
Comment (maelle) [reply vignettes/internals.Rmd:65]: `track_changes()`?

Comment (maelle) [reply vignettes/internals.Rmd:65]: `record_changes()`?

File: tests/testthat/_snaps/unbump_version/NEWS.md [hunk 8]
```diff
@@ -1,4 +1,4 @@
-<!-- NEWS.md is maintained by https://cynkra.github.io/fledge, do not edit -->
+<!-- NEWS.md is maintained by https://fledge.cynkra.com/, do not edit -->
```
Comment (krlmlr) [inline tests/testthat/_snaps/unbump_version/NEWS.md:1]: Where does that change come from? Can we move that to a separate PR?

Comment (maelle) [reply tests/testthat/_snaps/unbump_version/NEWS.md:1]: ooooh right I see what you mean now, let me change that.

Comment (maelle) [reply tests/testthat/_snaps/unbump_version/NEWS.md]: done

File: tests/testthat/_snaps/bump_version/NEWS2.md [hunk 9]
```diff
@@ -1,4 +1,4 @@
-<!-- NEWS.md is maintained by https://cynkra.github.io/fledge, do not edit -->
+<!-- NEWS.md is maintained by https://cynkra.github.io/fledge/, do not edit -->
```
Comment (krlmlr) [inline tests/testthat/_snaps/bump_version/NEWS2.md:1]: The terminal slash is new, and adds noise...

File: tests/testthat/_snaps/bump_version/NEWS2.md [hunk 10]
```diff
@@ -12,4 +12,5 @@
 
 # tea 0.0.0.9000
 
-* Added a `NEWS.md` file to track changes to the package.
+- Added a `NEWS.md` file to track changes to the package.
```
Comment (krlmlr) [inline tests/testthat/_snaps/bump_version/NEWS2.md:15]: Is this an artifact of the pandoc roundtrip?

File: NEWS.md [hunk 11]
```diff
@@ -29,6 +29,7 @@
 - Typo.
 
 
+
```
Comment (krlmlr) [inline NEWS.md:32]: I wonder if we can control the number of empty lines added here.

File: tests/testthat/_snaps/bump_version/NEWS2.md (hunk 10 above)
Comment (maelle) [reply tests/testthat/_snaps/bump_version/NEWS2.md:15]: yes

Comment (maelle) [reply tests/testthat/_snaps/bump_version/NEWS2.md:15]: let me do an experiment though

Comment (maelle) [reply tests/testthat/_snaps/bump_version/NEWS2.md:15]: now 100% sure that it is :-)

File: tests/testthat/_snaps/bump_version/NEWS2.md (hunk 9 above)
Comment (maelle) [reply tests/testthat/_snaps/bump_version/NEWS2.md]: fixed

**Comment by krlmlr** [comment]
Thanks. I don't have time to dive into the code, I trust you here. My only concern is that I don't really want to deal with unexpected changes to `NEWS.md` files when they are updated the next time. I'd rather preview them now, before merging this PR.

Is there a way to preview the roundtrip for a set of existing files, say, from https://krlmlr.github.io/actions-sync/ ?

**Comment by maelle** [comment]
> Is there a way to preview the roundtrip for a set of existing files, say, from [krlmlr.github.io/actions-sync](https://krlmlr.github.io/actions-sync/) ?

What would be a good list of repos to work with?

File: NEWS.md (hunk 11 above)
Comment (maelle) [reply NEWS.md]: it was a mistake, it should be two lines, now fixed.

**Comment by krlmlr** [comment]
The list of repos under https://krlmlr.github.io/actions-sync/ is good.

**Comment by maelle** [comment]
TODO 
- add "edited" in the tibble, for versions that have been edited.
- add "keep" by default for all versions for reading, remove as needed.
- write the "keep" ones using the raw lines.

I think this relies on the assumption that there are "edited" ones then "keep" ones, not some sort of alternating pattern.

File: DESCRIPTION [hunk 12]
```diff
@@ -1,6 +1,6 @@
 Package: fledge
 Title: Smoother Change Tracking and Versioning for R Packages
-Version: 0.1.0.9029
+Version: 0.2.0
```
Comment (maelle) [inline DESCRIPTION:3]: to be fixed

File: R/fledgling.R [hunk 13]
```diff
@@ -28,44 +28,132 @@ read_version <- function() {
...
+    h2 = grepl("##", news_lines[starts]), # TODO does not account for all syntaxes,
+    raw = map2_chr(starts, ends, ~ paste(news_lines[seq2(.x, .y)], collapse = "\n")),
+    news = split(news, seq_len(length(news)))
+  )
+
+  section_df$section_state <- "keep"
+
+  section_df$title <- names(news)
+
+  find_version <- function(text) {
```
Comment (maelle) [inline R/fledgling.R:92]: probably this should go into a function that takes a title and parses into the different component. A function that'd be tested.

**Comment by krlmlr** [comment]
pkgdown fails?

This is good, but I have second thoughts.

Can we implement this functionality without altering the external API? `bump_version()` would read, update, and write a fledgling object.

Let's take a look in a mini-hackathon.

File: R/commits.R [hunk 14]
```diff
@@ -2,10 +2,6 @@ with_repo <- function(code) {
   withr::with_dir(usethis::proj_get(), code)
 }
 
-local_repo <- function(.local_envir = caller_env()) {
```
Comment (maelle) [inline R/commits.R:5]: why has this been removed @krlmlr?

File: R/fledgling.R (hunk 13 above)
Comment (maelle) [reply R/fledgling.R:92]: see #610 :innocent:

**Comment by maelle** [comment]
@krlmlr will this be good to go after #610?

**Comment by krlmlr** [comment]
Thanks!
